python_modules/libraries/dagster-k8s/dagster_k8s/container_context.py

```diff
@@ -204,6 +215,218 @@ def get_environment_dict(self) -> Mapping[str, str]:
         return {env_var_tuple[0]: env_var_tuple[1] for env_var_tuple in parsed_env_var_tuples}
 
+    def _snake_case_allowed_fields(
+        self, only_allow_user_defined_k8s_config_fields: Mapping[str, Any]
+    ) -> Mapping[str, Any]:
+        result = {}
+
+        for key in only_allow_user_defined_k8s_config_fields:
+            if key == "namespace":
+                result[key] = only_allow_user_defined_k8s_config_fields[key]
+            else:
+                if key == "container_config":
+                    model_class = kubernetes.client.V1Container
+                elif key in {"job_metadata", "pod_template_spec_metadata"}:
+                    model_class = kubernetes.client.V1ObjectMeta
+                elif key == "pod_spec_config":
+                    model_class = kubernetes.client.V1PodSpec
+                elif key == "job_spec_config":
+                    model_class = kubernetes.client.V1JobSpec
+                else:
+                    raise Exception(f"Unexpected key in allowlist {key}")
+                result[key] = k8s_snake_case_keys(
+                    model_class, only_allow_user_defined_k8s_config_fields[key]
+                )
+        return result
+
+    def validate_user_k8s_config(
+        self,
+        only_allow_user_defined_k8s_config_fields: Optional[Mapping[str, Any]],
+        only_allow_user_defined_env_vars: Optional[Sequence[str]],
+    ) -> "K8sContainerContext":
+        used_fields = self._get_used_k8s_config_fields()
+
+        if only_allow_user_defined_k8s_config_fields is not None:
+            snake_case_allowlist = self._snake_case_allowed_fields(
+                only_allow_user_defined_k8s_config_fields
+            )
+
+            disallowed_fields = []
+
+            for key, used_fields_with_key in used_fields.items():
+                if isinstance(used_fields_with_key, set):
+                    for used_field in used_fields_with_key:
+                        if not snake_case_allowlist.get(key, {}).get(used_field):
+                            disallowed_fields.append(f"{key}.{used_field}")
+                else:
+                    check.invariant(isinstance(used_fields_with_key, bool))
+                    if used_fields_with_key and not only_allow_user_defined_k8s_config_fields.get(
+                        key
+                    ):
+                        disallowed_fields.append(key)
+
+            if disallowed_fields:
+                raise Exception(
+                    f"Attempted to create a pod with fields that violated the allowed list: {', '.join(disallowed_fields)}"
+                )
+
+        validated_container_context = self
+
+        if only_allow_user_defined_env_vars is not None:
+            validated_container_context = self._filter_user_defined_env_vars(
+                set(only_allow_user_defined_env_vars)
+            )
+
+        return validated_container_context
+
+    def _filter_user_defined_k8s_config_env_vars(
```

Comment:
easily bypassed by loading env from a configmap or secrets, provided the malicious user has access to provision these.

Reply:
That's true, but that's a different threat model than the one that we're preventing against here (and the other field in this PR gives you a way to prevent configmaps or secrets being set)